hfs_fit.py

```diff
@@ -0,0 +1,717 @@
+        relIntensities = []
+        for f in self.listF:
+            relIntensities.append(ri.RelIntensity(self.I, self.upperJ, self.lowerJ, f[0], f[1]))
+        self.relIntensities = np.array(relIntensities) #intensity ratios for each transition
+
+    def ParamsGuess(self, p = [0, 0, 0, 0, 0.150, 0.005, -1, -1, 0]):
+        '''
+        Sets all parameter guesses.
+        As, Bs, Gw and Lw are input in /cm.
+        p = [AUpperGuess, ALowerGuess, BUpperGuess, BLowerGuess, sGaussianGuess, gCauchyGuess, areaGuess, cogwnGuess, saturationGuess]
+        our results are usally doppler dominated so expect lorentzian width to be very small
+        CoG wavenumber and area guesses are calculated automatically based on the selected wavenumber range and line profiles
+        '''
+        #Obtain guesses for As and Bs
+        temp = np.array(p, dtype = 'float64')
+        temp[7] = (self.w * self.i).sum() / self.i.sum()
+        area = np.trapz(self.i, self.w)
+        temp[6] = area
+        self.paramsGuess = temp
+        #Guess B to be zero, since they tend to be small.
+        self.paramsGuess[2] = 0
+        self.paramsGuess[3] = 0
+        print('Current rms = ' + str(self.Residual(self.paramsGuess)))
+
+    def ApoFit(self, fit):
+        '''
+        Given a fit of Voigt functions sum, apodise the interferogram (phase included, using the complex amplitude)
+        of the fitted spectrum from where the observed interferogram first reaches zero.
+        this needs to be done as the mirrors moved a finite distance, apodising the interferogram and caused ringing of lines 
+        '''
+        I_fit = np.fft.fft(fit)
+        self.I_apo_fit = cp.copy(I_fit)
+        self.I_apo_fit[self.icut:-self.icut] = 0 #apodise real and complex part
+        S_apo_fit = np.fft.ifft(self.I_apo_fit)
+        return np.real(S_apo_fit)
+
+    def PlotApo(self):
+        '''
+        Shows line and guess line in interferogram domain, before and after apodisation of set self.icut
+        '''
+        #temp = self.FitLine(self.paramsGuess) #generates apodised and non-apodised fits
+        I_guess = np.fft.fft(self.noApoFit)
+        I_data = np.fft.fft(self.i)
+        plt.plot(self.w, np.abs(I_data), '-ro', label = 'data')
+        plt.plot(self.w, np.abs(I_guess), '-bo', label = 'fit')
+        plt.plot(self.w, np.abs(self.I_apo_fit), '-ko', label = 'apodised fit')
+        plt.legend()
+        plt.grid()
+        plt.show()
+
+    def Hold(self, b = [0, 0, 0, 0, 0, 0, 0, 0, 1]):
+        '''
+        Set which paramter to hold in self.paramGuess, 1 to hold, 0 to not hold.
+        same indices as self.paramsGuess
+        Can just turn each on and off in the fitting plot.
+        '''
+        self.boo = b
+
+    def Swing(self, swing = [1e-3, 1e-3, 1e-3, 1e-3, 1e-2, 1e-3, 5e-2, .1e-2, .1]):
+        '''
+        Sets Gaussian standard deviations for the jump distribution for each parameter.
+        As, Bs, Gw and Lw are input in /cm.
```

Comment:
could we use a dict rather than a list for these parameters, accidently using one instead of the other is a sure fire way to introduce a realllllly hard to find bug